### DISCUSSION TITLE
讨论下一个迭代的任务集

### DISCUSSION THREAD

**Discussion by cuixiaorui** [cuixueshe/earthworm#54] [General]
## 任务点

主要是分为 2 大模块
一个是技术方向  另外一个是功能方向
大家基于自己的使用感受来讨论讨论 目前还想要什么功能

### 技术
- [ ] client 单元测试
- [ ] client 集成测试
- [ ] api 的接口测试
- [ ] 前后端类型共享
- [ ] 新增用户和游客两个概念
- [ ] pinia store 的类型错误问题
- [ ] 关闭 auto import
	- 来源不清晰 维护成本过高  
- [ ] 在 git push 之前执行 test&lint&build 
	- 提前发现问题
- [ ] 封装 router 相关的逻辑
	- 目前关于路由的跳转都是基于写死的 path string 
	- 不好维护以及不便于复用
- [ ] 自动化部署脚本
### 功能
- [ ] 给予代码贡献的同学特殊的标识
- [ ] 背景音乐 - 沉浸式学习
- [ ] 错误的题目添加到后面的句子中
	- 保证把错误的句子搞正确后 才可以通关
	- 可参考 lingvist  
- [ ] 分享功能
	- 自动生成一张漂亮的分享图
		- 当前完成的课程是xx
		- 用时 xx
		- 已经连续坚持 xxx
		- pua 话术
		- 二维码
- [ ] 游客点击下一关后，注册完直接跳转到下一关
- [ ] 切换课程的时候需要显示 loading
- [ ] 提交句子 错误的话 不直接删除所有的内容
	- 需要提示出哪几个单词错了 
	- 提供更方便的修改交互方式
- [ ] 显示光标位置
	- 想要编辑其他单词时 需要移动光标 但是现在不知道光标的位置
	- 移动光标后  激活的位置也是错误的
- [ ] 对 win 友好的快捷键
- [ ] 实现打卡日历
	- 类似于 github 
	- ![image.png](https://images-1252602850.cos.ap-beijing.myqcloud.com/20240129132318.png)

### 团队
- [ ] 画图来表达出当前项目的所以概念(统一建模语言)
- [ ] apifox 的组织

**Comment by Nauxscript** [comment]
移动端适配或许可以考虑在 v2 或 v3 加入，不过适配的内容较多，交互，输入等

**Comment by hildxd** [comment]
- [ ] 给贡献用户设计类似 tag 那种标识

**Comment by hildxd** [comment]
- [ ] 完善后端 swagger 
- [ ] 使用 vue-query 替换 axios
- [ ] 在 scripts packages 里添加生成前端请求脚本

**Comment by hildxd** [comment]
- [ ] 去除nestjs @UncheckAuth 和 装饰器 改成@AuthUser 和 @User 装饰器

**Comment by Nauxscript** [comment]
彩蛋系统，如早上 10点24分 答题结束会触发特殊的效果，如 “早上好！爱学习的程序员！”

**Reply by myltx** [reply]
那这个是不是可以参考多邻国的模式，他在早上打卡的时候会有不同的提示，比如你早上八点打卡，他会说：”距离结束时间还有多少小时，提前完成了今日打卡，真棒“，这样的提示

**Reply by Nauxscript** [reply]
这种可能更像是一个固定的（一类）提示，不算彩蛋，但是可以作为一个功能点添加进去。
彩蛋的话更像是一些特殊条件才能触发的功能或内容，比如在愚人节，你无论你怎么答题，都会显示你今天没有打卡🤪（举个例子

**Reply by myltx** [reply]
那这个彩蛋还是好玩，固定类型的倒是可以确定一下该在什么样的时间去提示什么样的文案，感觉这个可以做了试试 🤔

**Reply by Nauxscript** [reply]
彩蛋系统与其说是一个“系统“倒不如说是一堆有趣的无伤大雅的玩笑 / 幽默 集合，在学习之余，附加点有趣的事物，会使学习这件事有趣一些。

**Reply by myltx** [reply]
悟了悟了，提高趣味性，然后调起学习的积极性  快乐加倍

**Comment by cuixiaorui** [comment]
在个人设置面板里面开一个 vim mode

可以使用 vim 的方式来编辑句子 

这个功能的性价比不高 但是我喜欢[旺柴]

**Reply by YoosonChan** [reply]
支持！

**Reply by baboon-king** [reply]
老板需求，可以往后放🥳

**Comment by Nauxscript** [comment]
- [ ] 设计个人主页面板页面 / 接口 / 个人信息表

**Reply by Nauxscript** [reply]
包含：
- [ ] 类 github 打卡日历
- [ ] 答题记录
- [ ] 个人主页整体页面
- [ ] 个人设置
- [ ] 头像 / 个人描述 （ 不确定

其他想法可以在这条以下回复

**Reply by seyeeL** [reply]
个人设置加一个点击下一题后是否默认播放声音

**Reply by Nauxscript** [reply]
有道理哈哈

**Reply by vampirewy** [reply]
引入英音/美音，并可以切换

**Reply by vampirewy** [reply]
如果有不同等级单词集(初中英语、高中英语、三级、四级，不过要看产品的定位)的话，可以设定每日单词学习数

**Reply by myltx** [reply]
那这个的基础上是不是可以统计你学过的单词，每个单词你学了多少遍，然后你输入的时候有哪些单词出现错误的情况更多，然后可以自己再针对性的去练习这块薄弱的内容。

**Comment by Wecle** [comment]
多个语法板块，比如：
1. 时态：一般时、进行时、完成时
2. 语态：主动、被动态
3. 语气：陈述、祈使、虚拟语气

此外的还可以有单词固定搭配练习、单词时态变形练习等等

**Reply by cuixiaorui** [reply]
后面的课程内容已经包含这部分的练习

**Reply by Wecle** [reply]
但是对于针对性学习还是比较难。有时候可能刷到都不一定知道这个是。

**Reply by fengstats** [reply]
要的就是这个效果，主打一个偷袭，不知不觉中把知识塞进脑子 [Doge]

**Comment by sadanspace** [comment]
想要个自定义内容的功能。
why：课程内容快刷完了，缺刷。
what：搞成什么样？还没想清楚。
用过类似的产品Anki，可以直接导入文件，基于文件生成对应练习。
后面可以做一个课程Hub，每个人都能做自己的练习专栏。
是不是后面可以直接输入一个网址，然后把这个网址的内容转换为练习专栏。以后想阅读啥文章，先练一遍，再去看的话会更容易。
how：第一版能够通过某种格式的文件导入进去来直接使用就行。

**Reply by Nauxscript** [reply]
这个崔哥有在考虑中了 @cuixiaorui 详细还要看实现计划

**Comment by Wecle** [comment]
打卡提醒：可以设定每日打卡时间，提醒要开始学习啦。

**Reply by Nauxscript** [reply]
怎么通知估计是个问题；邮件？
毕竟是个 web 端

**Reply by Wecle** [reply]
Web 有一个 [Notification API](https://developer.mozilla.org/en-US/docs/Web/API/Notification) 可以通过浏览器将消息推送到电脑通知栏、或者服务端推送邮件等等其他方式。具体更好的方式感觉可以调研一下。

**Comment by ryne6** [comment]
提问: 如何认领任务?

**Reply by hildxd** [reply]
这里讨论完后会开 issue 拆分任务，然后你去认领，给个截止时间就好

**Comment by vampirewy** [comment]
字有点小，看久了累pc端

**Reply by fengstats** [reply]
具体是哪个部分的字体小呢，如果是答题页面显示，可以善用浏览器的缩放功能哦~

![1706514358531](https://github.com/cuixueshe/earthworm/assets/48991003/fd246226-86e3-4492-8308-0d401e8f283c)

**Reply by vampirewy** [reply]
缩放功能是不会用的

**Reply by Jayden12138** [reply]
如果考虑功能上解决的话，可以在用户设置里添加对字号的设置，预设几种（大、中、小）或者自定义具体字号

**Comment by Nauxscript** [comment]
> * [ ]  给贡献用户设计类似 tag 那种标识

可以设计一个成就系统，类似 github 的 achievements

**Comment by myltx** [comment]
是否可以有一个积分系统？根据pr以及自己学习打卡等去获取积分

**Reply by vampirewy** [reply]
然后可以换实物哈哈哈，比如书这些

**Reply by Nauxscript** [reply]
哈哈哈好啊！快给项目拉点 sponsor！基金会！种子轮融资！搞快点！

**Reply by myltx** [reply]
哈哈 直接做大做强，那能不能直接拿自己的书去换积分，然后再拿积分去换其他人的东西 （感觉想的跑的有点远😂）

**Reply by Nauxscript** [reply]
你们想要我可以组织一下礼物交换😜

**Comment by hildxd** [comment]
- [ ] 写错单词不整句清空，保留有效数据
- [ ] 如果当前单词为空，按空格禁止跳转到 next input

**Reply by vampirewy** [reply]
个人觉得第一条可以不处理，加深一遍印象

**Reply by myltx** [reply]
第一个如果不处理后期长句感觉有点难顶，一下子清除一大句，如果打字快可能都没有注意到自己是哪个单词写错了

**Reply by Jayden12138** [reply]
对单词的拼写给一定的容错率，可以在做下一个前进行提示，某个单词的正确拼写应该是什么，拼错单词做重点标记

**Comment by myltx** [comment]
页面的样式风格都是自己去想然后画吗？有没有就是比较统一的风格要求然后大家在做的时候可以作为参考项去开发？

**Reply by Nauxscript** [reply]
目前没有 后续会尽量在 figma 中先做大概的设计再写页面；但是只能做大概的，毕竟目前都是程序员在参与😂

**Reply by myltx** [reply]
好嘞 比如个人中心那块我们是不是可以借鉴一下花瓣或者是其他设计网站的某些设计图，然后找几种比较类似的风格大家讨论一下挑选一个比较接近的风格去开发，这样也变相的有了简版设计 😂

**Reply by Nauxscript** [reply]
我们一般是直接借鉴现成的网站哈哈，当然风格上我们会选相近的网站，这个在细分开发任务的时候会注明

**Reply by myltx** [reply]
哈哈 那这个更合适 好嘞 等着任务了

**Reply by hildxd** [reply]
可以找我画 不过是另外的价钱

**Reply by Nauxscript** [reply]
你小子

**Reply by fengstats** [reply]
你小子

**Reply by myltx** [reply]
你小子

**Comment by yuanyangzi** [comment]
建议可以先进行英语水平测试，根据测试结果从建议的章节开始。（章节难度递增）

**Reply by Jayden12138** [reply]
那再增加个支持跳过当前章节，跳到指定章节

**Comment by baboon-king** [comment]
1. 启用 corepack ，并使用 `packageManager` 来固定包管理器。
```json
{
  "name": "earthworm",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "packageManager": "pnpm@8.15.5",
  "scripts": {}
}
```

2. 移除 engines 配置
该选项用于 当这个 project 是一个 npm 包时，用于限制安装者，并给予警告。 **earthworm** 是一个 WebApp 所以这个字段不用指定。

> 部分用户欲用来固定开发阶段的 Node 版本, 个人认为不恰当
> https://docs.npmjs.com/cli/v10/configuring-npm/package-json#engines

```json
"engines": {
  - "node": ">=20.11.0"
}
```

3. 使用 `.node-version` 来固定 Node 版本

> https://github.com/shadowspawn/node-version-usage

但是 `nvm` 还不支持。见 https://github.com/nvm-sh/nvm/issues/794

怎么办？我个人的习惯是：
1. 个人项目，我使用 `.node-version` with `fnm`
2. 协作项目/公司项目 我同时提供 `.node-version` 和 `.nvmrc` 来适应 nvm 的用户

对于 Earthworm
1. 同时提供 `.node-version` 和 `.nvmrc` 推荐
2. 只提供 `.node-version`，放弃 nvm 用户，不推荐
3. 只提供 `.nvmrc` ，`fnm` 可以兼容，推荐

**Comment by huoshidehua** [comment]
加入其他语言 比如日语

**Comment by xinshinfan** [comment]
这个项目用来练习外语太有效果了，最近在学日语，不知道大神会不会考虑加像日语这样的小语种，我可以整理提供素材💪🏻😄

**Reply by cuixiaorui** [reply]
可以呀

**Comment by refanbanzhang** [comment]
穿插错题或者提供一个入口集中练习错题？

**Comment by DanielWFH** [comment]
首先，我英语水平很垃圾（四级都过不了的水平）

**Reply by DanielWFH** [reply]
妈的不小心发出去了
我在学习的过程中其实更希望理解——输出，但是我发现我刷题刷久了我可以通过横杠的长短来猜单词，似乎我的大脑选择了一种最简单的回答问题的方法，避开了我本来应该要进行的思考，然后输出

**Reply by DanielWFH** [reply]
这个是我刷题刷久了的情况下发现的，也许是太累了，我他妈发现我的大脑在找捷径偷懒，笑死。

**Reply by fengstats** [reply]
@DanielWFH 你好，在设置中有个选项可以关闭显示单词长度，让每一个单词都是固定长度，应该可以缓解这个问题哈哈

![1713366467101](https://github.com/cuixueshe/earthworm/assets/48991003/36ab3787-a563-418e-ba38-04b02594453a)

**Reply by fengstats** [reply]
哈哈，正确情况，人的大脑会自己找偷懒的方法学习，后面熟练了之后可以试试切换 “听写” 模式，也是在设置里面

**Reply by DanielWFH** [reply]
wc nbnbnb
